sdk/typescript/src/providers/json-rpc-provider.ts

```diff
@@ -29,13 +30,57 @@ import {
   SuiObjectRef,
   getObjectReference,
   Coin,
+  SuiEventFilter,
+  SuiEventEnvelope,
+  SubscriptionId,
 } from '../types';
 import { SignatureScheme } from '../cryptography/publickey';
+import { Client as WsRpcClient} from 'rpc-websockets';
 
 const isNumber = (val: any): val is number => typeof val === 'number';
 const isAny = (_val: any): _val is any => true;
 
+const httpRegex = new RegExp('^http');
+const portRegex = new RegExp(':[0-9]{1,5}$');
+export const getWebsocketUrl = (httpUrl: string, port?: number): string => {
+  let wsUrl = httpUrl.replace(httpRegex, 'ws');
+  wsUrl = wsUrl.replace(portRegex, '');
+  return `${wsUrl}:${port ?? 9001}`;    // 9001 is full node websocket
+};
+
+enum ConnectionState {
+  NotConnected,
+  Connecting,
+  Connected
+}
+
+type JsonRpcMethodMessage<T> = {
+  jsonrpc: '2.0',
+  method: string,
+  params: T
+}
+
+type FilterSubHandler = {
+  id: SubscriptionId,
+  onMessage: (event: SuiEventEnvelope) => any,
+  filter: SuiEventFilter
+};
+
+type SubscriptionData = {
+  filter: SuiEventFilter,
+  onMessage: (event: SuiEventEnvelope) => any
+}
+
+export type SubscriptionEvent = { subscription: SubscriptionId, result: SuiEventEnvelope };
+
 export class JsonRpcProvider extends Provider {
+  protected wsClient: WsRpcClient;
+  protected wsConnectionState: ConnectionState = ConnectionState.NotConnected;
+  protected wsEndpoint: string;
+  protected wsTimeout: number = 0;
```

Comment:
I'd use `number | null`, so that it's a little more obvious when there is no timeout.